devcycle_python_sdk/api/config_client.py

```diff
@@ -0,0 +1,114 @@
+    APIClientError,
+    NotFoundError,
+    APIClientUnauthorizedError,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class ConfigAPIClient:
+    def __init__(self, sdk_key: str, options: DevCycleLocalOptions):
+        self.sdk_key = sdk_key
+        self.options = options
+        self.session = requests.Session()
+        self.session.headers = {
+            "Content-Type": "application/json",
+            "Accept": "application/json",
+        }
+        self.session.max_redirects = 0
+        self.max_config_retries = 2
+
+    def _config_file_url(self) -> str:
+        return join(self.options.config_CDN_URI, "v1", "server", self.sdk_key, ".json")
+
+    def get_config(self, config_etag: Optional[str] = None) -> Tuple[Optional[dict], Optional[str]]:
+        """
+        Get the config from the server. If the config_etag is provided, the server will only return the config if it
+        has changed since the last request. If the config hasn't changed, the server will return a 304 Not Modified
+        response.
+
+        :param config_etag: The etag of the last config request
+
+        :return: A tuple containing the config and the etag of the config. If the config hasn't changed since the last
+        request, the config will be None and the etag will be the same as the last request.
+        """
+        retries_remaining = self.max_config_retries
+        timeout = self.options.config_request_timeout_ms / 1000.0
+
+        url = self._config_file_url()
+
+        headers = {}
+        if config_etag:
+            headers["If-None-Match"] = config_etag
+
+        attempts = 1
+        while retries_remaining > 0:
+            request_error: Optional[Exception] = None
+            try:
+                res: requests.Response = self.session.request(
+                    "GET", url, params={}, timeout=timeout, headers=headers
+                )
+
+                if res.status_code == HTTPStatus.UNAUTHORIZED or res.status_code == HTTPStatus.FORBIDDEN:
+                    # Not a retryable error
+                    raise APIClientUnauthorizedError("Invalid SDK Key")
+                elif res.status_code == HTTPStatus.NOT_MODIFIED:
+                    # the config hasn't changed since the last request
+                    # don't return anything
+                    return None, config_etag
+                elif res.status_code == HTTPStatus.NOT_FOUND:
+                    # Not a retryable error
+                    raise NotFoundError(url)
+                elif HTTPStatus.BAD_REQUEST <= res.status_code < HTTPStatus.INTERNAL_SERVER_ERROR:
+                    # Not a retryable error
+                    raise APIClientError(f"Bad request: HTTP {res.status_code}")
+                elif res.status_code >= HTTPStatus.INTERNAL_SERVER_ERROR:
+                    # Retryable error
+                    request_error = APIClientError(
+                        f"Server error: HTTP {res.status_code}"
+                    )
```

Comment:
I think we need an `else` here for unexpected status codes.

Reply:
I don't think we do because the "else" case is that the request was successful in which case we fall out of the retry while loop and return.

This is the same thing you are doing in bucketing_client.py

Reply:
Yeah, good point. I'm guessing any other status codes (1xx, 3xx) are just going to be handled internally by requests anyways.